chain/rosetta-rpc/src/lib.rs

```diff
@@ -185,6 +185,80 @@ async fn network_options(
     }))
 }
 
+/// Get a block with `block_id`.
+/// Returns `Ok(Some(_))` if the block exists and is final.
+/// Returns `Ok(None)` if the block does not exist or is not final.
+async fn get_block_if_final(
+    block_id: &near_primitives::types::BlockReference,
+    view_client_addr: &Addr<ViewClientActor>,
+) -> Result<Option<near_primitives::views::BlockView>, models::Error> {
+    let mut is_query_by_height = false;
+    let mut is_query_final = false;
+    match block_id {
+        near_primitives::types::BlockReference::BlockId(
+            near_primitives::types::BlockId::Height(_),
+        ) => {
+            is_query_by_height = true;
+        }
+        near_primitives::types::BlockReference::Finality(
+            near_primitives::types::Finality::Final,
+        ) => {
+            is_query_final = true;
+        }
+        _ => {}
+    }
+    let block = match view_client_addr.send(near_client::GetBlock(block_id.clone())).await? {
+        Ok(block) => block,
+        Err(_) => return Ok(None),
+    };
+    if is_query_final {
+        return Ok(Some(block));
+    }
+    let final_block = match view_client_addr
+        .send(near_client::GetBlock(near_primitives::types::BlockReference::Finality(
+            near_primitives::types::Finality::Final,
+        )))
+        .await?
+    {
+        Ok(block) => block,
+        Err(_) => {
+            return Err(errors::ErrorKind::InternalError("final block not found".to_string()).into())
+        }
+    };
+    if block.header.height <= final_block.header.height {
```

Comment:
I’d negate the condition and make an early return:

```
    if block.header.height > final_block.header.height {
        return Ok(None);
    }
```

Excessive indentation and control flow makes code harder to read.